akka-http-core/src/main/scala/akka/http/impl/engine/http2/client/PersistentConnection.scala

```diff
@@ -55,20 +59,74 @@ private[http2] object PersistentConnection {
       object Unconnected extends State {
         override def onPush(): Unit = connect()
         override def onPull(): Unit =
-          if (!hasBeenPulled(requestIn)) // requestIn might already have been pulled when we failed and went back to Unconnected
+          if (!isAvailable(requestIn) && !hasBeenPulled(requestIn)) // requestIn might already have been pulled when we failed and went back to Unconnected
             pull(requestIn)
-
       }
 
       def connect(): Unit = {
         val requestOut = new SubSourceOutlet[HttpRequest]("PersistentConnection.requestOut")
         val responseIn = new SubSinkInlet[HttpResponse]("PersistentConnection.responseIn")
+        val connection = Promise[OutgoingConnection]
+
+        become(new Connecting(connection.future, requestOut, responseIn))
+
+        connection.completeWith(Source.fromGraph(requestOut.source)
+          .viaMat(connectionFlow)(Keep.right)
+          .toMat(responseIn.sink)(Keep.left)
+          .run()(subFusingMaterializer))
+      }
+
+      class Connecting(
+        connected:  Future[OutgoingConnection],
+        requestOut: SubSourceOutlet[HttpRequest],
+        responseIn: SubSinkInlet[HttpResponse]
+      ) extends State {
+        implicit val ec = materializer.executionContext
+        connected.onComplete {
+          case Success(_) =>
+            onConnected.invoke(())
+          case Failure(_) =>
+            onFailed.invoke(())
+        }
```

Comment:
Probably does not matter much but this could be done on parasitic/samethread since async callback invoke should never fail.

Reply:
makes sense, done